cosmoz-default-tree.html

```diff
@@ -15,142 +27,236 @@
+	 * @param {object} nodes [this._roots] (The nodes the search should be based on.)
+	*/
+	Cosmoz.DefaultTree.prototype._searchNodes = function (propertyValue, options, nodes) {
+		var results = [];
 
-		if (propertyName === 'pathLocator') {
-			return this._getPathByPathLocator(propertyValue);
-		}
+		// Defaults
+		nodes = nodes || this._roots;
 
-		var node = this._searchAllRoots(propertyName, propertyValue);
-		if (node !== undefined) {
-			return this._getPathByPathLocator(node.pathLocator);
-		}
-	};
+		nodes.some(function (node) {
+			results = results.concat(this.search(node, propertyValue, options));
+			return options.firstHitOnly && results.length > 0;
+		}, this);
 
-	Cosmoz.DefaultTree.prototype.getChildren = function (node) {
-		if (node === undefined) {
-			return;
-		}
-		return _objectValues(node.children);
+		return results;
 	};
 
-	Cosmoz.DefaultTree.prototype.getProperty = function (node, propertyName) {
-		if (node === undefined || propertyName === undefined) {
-			return;
+	/**
+	 * Returns the node of a given path.
+	 * @return {object}
+	 * @param {string} pathLocator (The string which describes the path. e.g. "1.2.9")
+	 * @param {object} nodeObj [this._treeData] (The object the search should be based on.)
+	 * @param {string} separatorSign ["."] (The string which separates the path. e.g ".")
+	*/
+	Cosmoz.DefaultTree.prototype.getNodeByPathLocator = function (pathLocator, nodeObj, separatorSign) {
+		if (!pathLocator) {
+			return this._roots;
 		}
 
-		return node[propertyName];
-	};
+		// Defaults
+		nodeObj = nodeObj || this._treeData;
+		separatorSign = separatorSign || '.';
 
-	Cosmoz.DefaultTree.prototype._searchAllRoots = function (propertyName, propertyValue) {
-		var i,
-			node;
+		var	pathNodes = this.getPathNodes(pathLocator, nodeObj, separatorSign);
+		return pathNodes && pathNodes.pop();
+	};
 
-		for (i = 0; i < this._roots.length; i+=1) {
-			node = this.search(this._roots[i], propertyName, propertyValue);
-			if (node !== undefined) {
-				return node;
-			}
+	/**
+	 * Returns the nodes on a given path.
+	 * @return {array}
+	 * @param {string} pathLocator (The string which describes the path. e.g. "1.2.9")
+	 * @param {object} nodeObj [this._treeData] (The object the search should be based on.)
+	 * @param {string} separatorSign ["."] (The string which separates the path.)
+	*/
+	Cosmoz.DefaultTree.prototype.getPathNodes = function (pathLocator, nodeObj, separatorSign) {
+		if (!pathLocator) {
+			return this._roots;
 		}
-	};
 
-	Cosmoz.DefaultTree.prototype._getNodeByPathLocator = function (pathLocator) {
-		var	node,
-			path,
-			i;
+		// Defaults
+		separatorSign = separatorSign || '.';
+
+		var path = pathLocator.split(separatorSign),
+			pathSegment = nodeObj || this._treeData,
+			nodes = [];
 
-		for (i = 0; i < this._roots.length; i+=1) {
-			node = this._roots[i];
-			if (node.pathLocator === pathLocator) {
-				return node;
+		// Get the nodes on the path.
+		path.some(function (nodeKey) {
```

Comment:
Well this is the behavior I would expect with valid path `1.2.3`:

1. Correct path - `1.2` = `[1, 2]`
2. Missing start - `0.1.2` = `[undefined, 1, 2]`
Does not work with your implementation.
Should also probably be trimmed to `[1, 2]`
3. Missing middle - `1.4.2` = `[1, undefined, undefined]`
This should not become `[1, undefined, 2]` as we had a scenario earlier.
4. Missing end - `1.2.3.3` = `[1, 2, 3, undefined]`
This should also not become `[1, 2, 3, 3]` where the last node gets inserted twice as we had in a scenario.

But as you were doing earlier, we might just want to invalidate the whole path if we get `undefined` values somewhere else than in the beginning.
It's an easier approach right now and any additions to that can come later.

So I'm thinking the behavior should be something like this:

```js
	var validPath = path.every(function (nodeKey) {
		if (!pathSegment || !pathSegment[nodeKey]) {
			return nodes.length === 0;
		}
		var node = pathSegment[nodeKey];
		pathSegment = node[childProperty];
		nodes.push(node);
		return true;
	}, this);

	if (!validPath) {
		return null;
	}

	// corner case - no hit at all
	if (validPath && nodes.length === 0 && path.length > 0) {
		return null;
	}

	// filter out undefined items from the start
	return nodes.filter(function (node) {
		return node !== undefined;
	});
```

Also I think this should have quite a few tests :)